File: summary/dart/21_local_search.dart
List<int> localSearchAlgorithm(List<int> arr, int Function(List<int>) costFunction) {
  int n = arr.length;
  List<int> bestSolution = List.from(arr);
  int bestCost = costFunction(bestSolution);

  bool improved;
  do {
    improved = false;
    for (int i = 0; i < n; i++) {
      for (int j = i + 1; j < n; j++) {
        List<int> currentSolution = List.from(bestSolution);
        int temp = currentSolution[i];
        currentSolution[i] = currentSolution[j];
        currentSolution[j] = temp;

        int currentCost = costFunction(currentSolution);
        if (currentCost < bestCost) {
          bestSolution = List.from(currentSolution);
          bestCost = currentCost;
          improved = true;
        }
      }
    }
  } while (improved);

  return bestSolution;
}

void main() {
  // Example usage:
  List<int> inputArray = [5, 3, 8, 6, 2];
  List<int> bestArray = localSearchAlgorithm(inputArray, (arr) => arr.fold(0, (sum, num) => sum + num * num));
  print("Optimized Array: $bestArray");
}

File: summary/rust/21_local_search.rs
fn local_search_algorithm<F>(arr: &mut [i32], cost_function: F) -> Vec<i32>
where
    F: Fn(&[i32]) -> i32,
{
    let n = arr.len();
    let mut best_solution = arr.to_vec();
    let mut best_cost = cost_function(&best_solution);

    loop {
        let mut current_solution = best_solution.clone();
        let mut improved = false;

        for i in 0..n {
            for j in i + 1..n {
                current_solution.swap(i, j);
                let current_cost = cost_function(&current_solution);
                if current_cost < best_cost {
                    best_solution = current_solution.clone();
                    best_cost = current_cost;
                    improved = true;
                }
                current_solution.swap(i, j);
            }
        }

        if !improved {
            break;
        }
    }

    best_solution
}

// Example usage:
fn main() {
    let mut input_array = vec![4, 3, 2, 1];
    let best_array = local_search_algorithm(&mut input_array, |arr| arr.iter().map(|&num| num * num).sum());
    println!("Optimized Array: {:?}", best_array);
}
